sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotUseCollectionInItsOwnMethodCalls.cs

```diff
@@ -0,0 +1,212 @@
+            {
+                return AlwaysSameCollectionMessage;
+            }
+
+            if (IsISetExceptWithImplementation(methodSymbol) ||
+                IsISetSymmetricExceptWithImplementation(methodSymbol))
+            {
+                return AlwaysEmptyCollectionMessage;
+            }
+
+            if (IsISetIsSubsetOfImplementation(methodSymbol) ||
+                IsISetIsSupersetOfImplementation(methodSymbol) ||
+                IsISetOverlapsImplementation(methodSymbol) ||
+                IsISetSetEqualsImplementation(methodSymbol))
+            {
+                return AlwaysTrueMessage;
+            }
+
+            if (IsISetIsProperSubsetOfImplementation(methodSymbol) ||
+                IsISetIsProperSupersetOfImplementation(methodSymbol))
+            {
+                return AlwaysFalseMessage;
+            }
+
+            return null;
+        }
+
+        private static bool IsISetUnionWithImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "UnionWith");
+        private static bool IsISetExceptWithImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "ExceptWith");
+        private static bool IsISetIntersectWithImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "IntersectWith");
+        private static bool IsISetIsProperSubsetOfImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "IsProperSubsetOf");
+        private static bool IsISetIsProperSupersetOfImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "IsProperSupersetOf");
+        private static bool IsISetIsSubsetOfImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "IsSubsetOf");
+        private static bool IsISetIsSupersetOfImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "IsSupersetOf");
+        private static bool IsISetOverlapsImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "Overlaps");
+        private static bool IsISetSetEqualsImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "SetEquals");
+        private static bool IsISetSymmetricExceptWithImplementation(IMethodSymbol methodSymbol) =>
+            IsISetMethodImplementation(methodSymbol, "SymmetricExceptWith");
+
+        private static bool IsISetMethodImplementation(IMethodSymbol methodSymbol, string methodName)
+        {
+            return methodSymbol != null &&
+                methodSymbol.Name == methodName &&
+                methodSymbol.MethodKind == MethodKind.Ordinary &&
+                methodSymbol.Parameters.Length == 1 &&
+                methodSymbol.ContainingType.Implements(KnownType.System_Collections_Generic_ISet_T);
```

Comment:
It is rare, but you could also check for explicit interface implementations (see KnownMethods for examples)